packages/lodestar-beacon-state-transition/src/fast/util/epochContext.ts

```diff
@@ -55,12 +59,9 @@ export class EpochContext {
 
   public copy(): EpochContext {
     const ctx = new EpochContext(this.config);
-    // full copy of pubkeys, this can mutate
-    ctx.pubkey2index = new PubkeyIndexMap();
-    for(const entry of this.pubkey2index.entries()) {
-      ctx.pubkey2index.set(fromHexString(entry[0] as unknown as string), entry[1]);
-    }
-    ctx.index2pubkey = [...this.index2pubkey];
+    // warning: pubkey cache is not copied, it is shared, as eth1 is not expected to reorder validators.
```

Comment:
We copied it because state transition would modify pubkey cache for all previous epoch contexts.

Like you have context with 8 validators for epoch 1, and when transition to epoch 2 you copy this and for example add 3 and remove 2 validators, you will update epoch context for epoch 1 so we would have bad transition in case of reorg from epoch 1 block

Reply:
The pubkey cache won't reorg unless the deposit order reorgs, for now in phase 0 at least. So it's a case of only adding pubkeys, never overwriting/removing anything. The pubkeys that are not yet supposed to be in the cache w.r.t. the state can be ignored on the usage side by checking the validator index against the current registry size.
Now, the difficult part is avoiding bad reorgs, when e.g. an attacker manages to opportunistically fabricate a bad chain with censored votes.
If the cache is only used for finalized pubkeys it is much less of a problem, but more difficult.

Keeping the copy, but using an immutable map structure for cheap copies sounds like a clean better approach.